File: clients/kotlin/src/test/kotlin/net/eternaltwin/auth/AuthContextTest.kt
package net.eternaltwin.auth;

import net.eternaltwin.SerializationTestItem
import net.eternaltwin.oauth.OauthClientDisplayName
import net.eternaltwin.oauth.OauthClientId
import net.eternaltwin.oauth.OauthClientKey
import net.eternaltwin.oauth.ShortOauthClient
import net.eternaltwin.user.*
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream
import kotlin.test.assertEquals

class AuthContextTest {
  companion object {
    @JvmStatic
    fun fromJsonString(): Stream<SerializationTestItem<AuthContext>> =
      SerializationTestItem.streamFromTestDir(
        "auth/auth-context",
        mapOf(
          "demurgos" to AuthContext.User(
            UserAuthContext(
              scope = AuthScope.Default,
              user = ShortUser(
                id = UserId("9f310484-963b-446b-af69-797feec6813f"),
                displayName = UserDisplayNameVersions(UserDisplayNameVersion(UserDisplayName("Demurgos"))),
              ),
              isAdministrator = true
            )
          ),
          "eternalfest-demurgos" to AuthContext.AccessToken(
            AccessTokenAuthContext(
              scope = AuthScope.Default,
              client = ShortOauthClient(
                id = OauthClientId("d19e61a3-83d3-410f-84ec-49aaab841559"),
                key = OauthClientKey("eternalfest@clients"),
                displayName = OauthClientDisplayName("Eternalfest"),
              ),
              user = ShortUser(
                id = UserId("9f310484-963b-446b-af69-797feec6813f"),
                displayName = UserDisplayNameVersions(UserDisplayNameVersion(UserDisplayName("Demurgos"))),
              )
            )
          ),
          "guest" to AuthContext.Guest(
            GuestAuthContext(
              scope = AuthScope.Default,
            )
          ),
        ),
      )
  }

  @ParameterizedTest
  @MethodSource
  fun fromJsonString(item: SerializationTestItem<AuthContext>) {
    val actual = AuthContext.fromJsonString(item.jsonString)
    assertEquals(item.value, actual)
  }
}

File: clients/php/test/Auth/AuthContextTest.php
<?php declare(strict_types=1);

namespace Eternaltwin\Test\Auth;

use Eternaltwin\Auth\AccessTokenAuthContext;
use Eternaltwin\Auth\AuthScope;
use Eternaltwin\Auth\AuthContext;
use Eternaltwin\Auth\GuestAuthContext;
use Eternaltwin\Auth\UserAuthContext;
use Eternaltwin\Oauth\OauthClientDisplayName;
use Eternaltwin\Oauth\OauthClientId;
use Eternaltwin\Oauth\OauthClientKey;
use Eternaltwin\Oauth\ShortOauthClient;
use Eternaltwin\Test\SerializationTestItem;
use Eternaltwin\User\ShortUser;
use Eternaltwin\User\UserDisplayName;
use Eternaltwin\User\UserDisplayNameVersion;
use Eternaltwin\User\UserDisplayNameVersions;
use Eternaltwin\User\UserId;
use PHPUnit\Framework\TestCase;

final class AuthContextTest extends TestCase {
  /**
   * @dataProvider provideFromJson
   * @param SerializationTestItem $item
   */
  public function testFromJson(SerializationTestItem $item): void {
    $actual = AuthContext::fromJson($item->getJson());
    $this->assertEquals($item->getValue(), $actual);
  }

  public function provideFromJson(): array {
    return SerializationTestItem::fromTestDir(
      "core/auth/auth-context",
      [
        "demurgos" => new UserAuthContext(
          AuthScope::Default(),
          new ShortUser(
            UserId::fromString("9f310484-963b-446b-af69-797feec6813f"),
            new UserDisplayNameVersions(new UserDisplayNameVersion(new UserDisplayName("Demurgos")), []),
          ),
          true,
        ),
        "eternalfest-demurgos" => new AccessTokenAuthContext(
          AuthScope::Default(),
          new ShortOauthClient(
            OauthClientId::fromString("d19e61a3-83d3-410f-84ec-49aaab841559"),
            new OauthClientKey("eternalfest@clients"),
            new OauthClientDisplayName("Eternalfest"),
          ),
          new ShortUser(
            UserId::fromString("9f310484-963b-446b-af69-797feec6813f"),
            new UserDisplayNameVersions(new UserDisplayNameVersion(new UserDisplayName("Demurgos")), []),
          ),
        ),
        "guest" => new GuestAuthContext(
          AuthScope::Default(),
        ),
      ],
    );
  }
}
